Remove unwraps from tls13 handle_client_hello

ab-cgh opened the pull request:
This one involved a bit of binding refactoring. I think it matches the intent better though, since [this comment](https://github.com/ctz/rustls/blob/main/rustls/src/server/tls13.rs#L590) mentions "key share" and not "group".

djc commented on rustls/src/server/tls13.rs:
```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
+                // We don't have a suitable key share.  Choose a suitable group and
+                // send a HelloRetryRequest.
+                let retry_group_maybe = supported_groups
                 .iter()
                 .find(|group| groups_ext.contains(group))
                 .cloned();
```
I think these will now need an extra bit of indentation.

ab-cgh replied:
Good catch, thanks. Addressed.

djc commented on rustls/src/server/tls13.rs:
```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
             .iter()
-            .find(|group| share_groups.contains(group))
-            .cloned();
+            .find(|group| share_groups.contains(group)) // Find the chosen group
+            .and_then(|g| shares_ext.iter().find(|k| k.group == *g)); // Then its share
```
I'd prefer using `group` over `g` and `share` over `k` -- this code is dense enough as it is...

ab-cgh replied:
Names changed, good point.

djc reviewed (commented):
Some style nits here. There's actually also a behavioral change where we're now treating more types of incompatibility with the same error handling. For that, I'd like Brian or Joe to have a look as well.

briansmith commented on rustls/src/server/tls13.rs:
```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
-            // We don't have a suitable key share.  Choose a suitable group and
-            // send a HelloRetryRequest.
-            let retry_group_maybe = supported_groups
+        let chosen_share =  match chosen_share {
```
`.unwrap_or_else`?

briansmith replied:
Sorry, never mind that comment.

ab-cgh commented:
I didn't see the behavior change originally. I had thought any group, if it existed, would also have a share associated with it. If it didn't have one, it would of hit the `.unwrap()` later down. AFAIK this change keeps the original behavior of if no group exists, then it takes the route of sending a `HelloRetryRequest`.

```
                            Original:
try to get group ------------------------------------------------------------------
                                                                                  v 
    Modified:                    | --> None --> skip group->share search -------> |                                  | --> None --> send HelloRetryRequest --> ...
try to get group --> iterator ----                                                | --> match on groups existence -->|          Original:
                                 | --> Some --> search for share in group too --> |                                  | --> Some --> unwrap on group->share search
                                                                                                                                Modified:
                                                                                                                            Some --> no unwrap needed
```

briansmith reviewed (commented):
PTAL at a slightly different alternative: PR #489.

ab-cgh commented:
I like the implementation in #489 as well. I'm happy to close this PR in favor if it if desired. I have no strong feelings either way.

briansmith commented on rustls/src/server/tls13.rs:
```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
-            self.handshake
-                .transcript
-                .add_message(chm);
+            .find(|group| share_groups.contains(group)) // Find the chosen group
```
I think you can use `find_map(|group| shares_ext.iter().find(|share| share.group == *group)` here to avoid doing two lookups.

ab-cgh replied:
Good idea, and change applied. This worked nicely.

briansmith commented on rustls/src/server/tls13.rs:
```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
-                return Ok(self.into_expect_retried_client_hello());
-            }
+                if let Some(group) = retry_group_maybe {
+                    if self.done_retry {
```
> It would be clearer to check self.done_retry error out before we try to calculate retry_group_maybe.

We should probably do that reorganization to make it clearer, but my suggestion is out of scope for this PR.

briansmith commented:
I close my other PR because I noticed it is subtly wrong, and because it was trying to do things that are out of scope of this. I think the `find_map` suggestion above is the only thing I would change.

briansmith commented:
> I didn't see the behavior change originally. I had thought any group, if it existed, would also have a share associated with it.

I'm not understanding. The `share_groups` collection is derived from the `share_exts` collection so it is indeed true that every item in `shared_groups` exists in `share_exts`. If you take my `find_map` suggestion then I think you can eliminate `share_groups`.

ab-cgh commented:
Your suggestion did get rid of `share_groups`, but that explanation was an attempt at showing that the final logic did not change with this PR from what it originally was due to removing `chosen_group`.

briansmith commented:
LGTM. I think this is much clearer than the original code.

djc commented:
Yup, this looks good, thanks!
